fontc_crater/src/ci.rs

```diff
@@ -0,0 +1,118 @@
+//! Running in CI.
+//!
+//! This handles the whole CI workflow.
+//!
+//! Unlike a normal run, this is expecting to have preexisting results, and to
+//! generate a fuller report that includes comparison with past runs.
+
+use std::path::{Path, PathBuf};
+
+use chrono::{DateTime, Utc};
+use google_fonts_sources::RepoInfo;
+
+use crate::{
+    args::CiArgs,
+    error::Error,
+    ttx_diff_runner::{DiffError, DiffOutput},
+    Results,
+};
+
+mod html;
+
+static SUMMARY_FILE: &str = "summary.json";
+
+type DiffResults = Results<DiffOutput, DiffError>;
+
+/// A summary of a single CI run
+
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
+struct RunSummary {
+    began: DateTime<Utc>,
+    finished: DateTime<Utc>,
+    fontc_rev: String,
+    results_file: PathBuf,
+    // the name of the file listing targets used by this run.
+    // it is intended that when this list is updated, the filename is changed.
+    input_file: PathBuf,
+    stats: super::ttx_diff_runner::Summary,
+}
+
+impl RunSummary {
+    fn try_load_results(&self, target_dir: &Path) -> Option<Result<DiffResults, Error>> {
+        let report_path = target_dir.join(&self.results_file);
+        if !report_path.exists() {
+            return None;
+        }
+        Some(super::try_read_json(&report_path))
+    }
+}
+
+pub(super) fn run_ci(args: &CiArgs) -> Result<(), Error> {
+    if !args.html_only {
+        super::ttx_diff_runner::assert_can_run_script();
+        run_crater_and_save_results(args)?;
+    }
+    html::generate(&args.out_dir)?;
+    // now we want to generate an html report, based on this info.
+    Ok(())
+}
+
+fn run_crater_and_save_results(args: &CiArgs) -> Result<(), Error> {
+    if !args.out_dir.exists() {
+        super::try_create_dir(&args.out_dir)?;
+    }
+
+    let summary_file = args.out_dir.join(SUMMARY_FILE);
+    let mut prev_runs: Vec<RunSummary> = if summary_file.exists() {
+        super::try_read_json(&summary_file)?
+    } else {
+        Default::default()
+    };
+
+    // todo: fontc_repo should be checked out by us, and have a known path
+    let fontc_rev = super::get_git_rev(None).unwrap();
+    if prev_runs
+        .last()
+        .map(|prev| prev.fontc_rev == fontc_rev)
+        .unwrap_or(false)
+    {
+        // eventually we will want an argument to force rerunning, which we'll
+        // use when we update fontmake or the input data set
+        eprintln!("fontc rev is unchange from last run, skipping");
+        std::process::exit(0);
```

Comment:
I really really dislike functions that kill my process, perhaps we an return a specific error? Also ... just because fontc didn't change doesn't mean fontmake or the sources didn't?

Reply:
This isn't an error condition, this just means we have no more work to do, and we're exiting happily.